File: scripts/convert-translations-to-ts.mjs
#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';

const repoRoot = path.resolve(process.argv[2] ?? path.join(process.cwd(), '..'));
const translationsRoot = path.join(repoRoot, 'src', 'translations');
const outputDir = path.join(repoRoot, 'escc', 'src', 'i18n');

const namespaces = ['common', 'groups', 'hints', 'log', 'settings'];

function xmlEscape(value) {
  return value
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&apos;');
}

function readJson(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  return JSON.parse(content);
}

function localeToQt(locale) {
  return locale.toLowerCase().replace('-', '_');
}

function buildTsForLocale(locale, sourcesByNamespace) {
  const localeDir = path.join(translationsRoot, locale);
  const localeCode = localeToQt(locale);
  const lines = [];

  lines.push('<?xml version="1.0" encoding="utf-8"?>');
  lines.push(`<TS version="2.1" language="${xmlEscape(localeCode)}">`);

  for (const ns of namespaces) {
    const localeFile = path.join(localeDir, `${ns}.json`);
    if (!fs.existsSync(localeFile)) {
      continue;
    }

    const localeJson = readJson(localeFile);
    const sourceJson = sourcesByNamespace.get(ns) ?? {};

    lines.push('  <context>');
    lines.push(`    <name>${xmlEscape(`esc.${ns}`)}</name>`);

    const keys = Object.keys(sourceJson);
    for (const key of keys) {
      const sourceText = String(sourceJson[key] ?? '');
      const translatedText = localeJson[key] !== undefined ? String(localeJson[key]) : sourceText;

      lines.push('    <message>');
      lines.push(`      <source>${xmlEscape(sourceText)}</source>`);
      lines.push(`      <comment>${xmlEscape(`${ns}.${key}`)}</comment>`);
      lines.push(`      <translation>${xmlEscape(translatedText)}</translation>`);
      lines.push('    </message>');
    }

    lines.push('  </context>');
  }

  lines.push('</TS>');
  lines.push('');
  return lines.join('\n');
}

function main() {
  if (!fs.existsSync(translationsRoot)) {
    throw new Error(`Translations folder not found: ${translationsRoot}`);
  }

  fs.mkdirSync(outputDir, { recursive: true });

  const locales = fs
    .readdirSync(translationsRoot, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name)
    .sort((a, b) => a.localeCompare(b));

  if (!locales.includes('en')) {
    throw new Error('Base locale "en" is required for source strings.');
  }

  const sourcesByNamespace = new Map();
  for (const ns of namespaces) {
    const sourceFile = path.join(translationsRoot, 'en', `${ns}.json`);
    if (!fs.existsSync(sourceFile)) {
      throw new Error(`Missing base source file: ${sourceFile}`);
    }
    sourcesByNamespace.set(ns, readJson(sourceFile));
  }

  for (const locale of locales) {
    const tsContent = buildTsForLocale(locale, sourcesByNamespace);
    const outputFile = path.join(outputDir, `escc_${localeToQt(locale)}.ts`);
    fs.writeFileSync(outputFile, tsContent, 'utf8');
  }

  console.log(`Generated TS files for ${locales.length} locales in ${outputDir}`);
}

main();

File: scripts/convert-translations-to-ts.py
#!/usr/bin/env python3

import json
from pathlib import Path
from xml.sax.saxutils import escape

ROOT = Path(__file__).resolve().parents[2]
TRANSLATIONS_ROOT = ROOT / "src" / "translations"
OUTPUT_DIR = ROOT / "escc" / "src" / "i18n"
NAMESPACES = ("common", "groups", "hints", "log", "settings")


def locale_to_qt(locale: str) -> str:
    return locale.lower().replace("-", "_")


def load_json(file_path: Path):
    with file_path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def main() -> None:
    if not TRANSLATIONS_ROOT.exists():
        raise RuntimeError(f"Translations folder not found: {TRANSLATIONS_ROOT}")

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    locales = sorted(path.name for path in TRANSLATIONS_ROOT.iterdir() if path.is_dir())
    if "en" not in locales:
        raise RuntimeError("Base locale 'en' is required")

    sources = {ns: load_json(TRANSLATIONS_ROOT / "en" / f"{ns}.json") for ns in NAMESPACES}

    for locale in locales:
        locale_dir = TRANSLATIONS_ROOT / locale
        lines = [
            '<?xml version="1.0" encoding="utf-8"?>',
            f'<TS version="2.1" language="{escape(locale_to_qt(locale))}">',
        ]

        for namespace in NAMESPACES:
            locale_file = locale_dir / f"{namespace}.json"
            if not locale_file.exists():
                continue

            translated = load_json(locale_file)
            lines.append("  <context>")
            lines.append(f"    <name>{escape(f'esc.{namespace}')}</name>")

            for key, source_text in sources[namespace].items():
                source_value = str(source_text)
                translation_value = str(translated.get(key, source_value))
                lines.append("    <message>")
                lines.append(f"      <source>{escape(source_value)}</source>")
                lines.append(f"      <comment>{escape(f'{namespace}.{key}')}</comment>")
                lines.append(f"      <translation>{escape(translation_value)}</translation>")
                lines.append("    </message>")

            lines.append("  </context>")

        lines.append("</TS>")
        lines.append("")
        output_file = OUTPUT_DIR / f"escc_{locale_to_qt(locale)}.ts"
        output_file.write_text("\n".join(lines), encoding="utf-8")

    print(f"Generated {len(locales)} TS files into {OUTPUT_DIR}")


if __name__ == "__main__":
    main()
